include/fiction/algorithms/simulation/sidb/assess_physical_population_stability.hpp

```diff
@@ -0,0 +1,377 @@
+
+template <typename Lyt>
+class assess_physical_population_stability_impl
+{
+  public:
+    /**
+     * Constructor for assess_physical_population_stability_impl.
+     *
+     * @param lyt SiDB layout.
+     * @param parameters The simulation parameters used for the assessment.
+     */
+    assess_physical_population_stability_impl(const Lyt&                                         lyt,
+                                              const assess_physical_population_stability_params& parameters) :
+            layout{lyt},
+            params{parameters}
+    {}
+
+    /**
+     * Runs a population stability assessment for a given SiDB layout using the provided simulation parameters.
+     * This function determines the minimum electrostatic potential required for charge state transitions within the
+     * layout and identifies the corresponding critical SiDB along with the type of charge state transition.
+     *
+     * @return A vector of population stability information structures, where each element represents a charge
+     * distribution in ascending energy order. Each structure contains details about the critical SiDB, the type of
+     * charge state transition, and the minimum electrostatic potential required for the charge transition.
+     */
+    [[nodiscard]] std::vector<population_stability_information<Lyt>> run() noexcept
+    {
+        const quickexact_params<Lyt> quickexact_parameters{params.physical_parameters};
+        const auto                   simulation_results = quickexact(layout, quickexact_parameters);
+        const auto energy_and_unique_charge_index       = collect_energy_and_charge_index(simulation_results);
+
+        std::vector<population_stability_information<Lyt>> popstability_information{};
+        popstability_information.reserve(simulation_results.charge_distributions.size());
+
+        // Access the unique indices
+        for (const auto& energy_and_index : energy_and_unique_charge_index)
+        {
+            const auto it = std::find_if(
+                simulation_results.charge_distributions.begin(), simulation_results.charge_distributions.end(),
+                [&](const charge_distribution_surface<Lyt>& charge_lyt)
+                {
+                    // Compare with the first element of the pair returned by get_charge_index_and_base()
+                    return charge_lyt.get_charge_index_and_base().first == energy_and_index.charge_index;
+                });
+
+            if (it == simulation_results.charge_distributions.end())
+            {
+                continue;
+            }
+
+            const auto                            charge_lyt = *it;
+            population_stability_information<Lyt> population_stability_info{};
+            population_stability_info.minimum_potential_difference_to_transition =
+                std::numeric_limits<double>::infinity();
+
+            charge_lyt.foreach_cell(
+                [this, &it, &population_stability_info](const auto& c)
```

Comment:
```suggestion
                [this, &charge_lyt, &population_stability_info](const auto& c)
```
Otherwise, this will not compile.